In [ ]:
import requests
from bs4 import BeautifulSoup
import pandas as pd


In [1]:
import pandas as pd

In [2]:
matches = pd.read_csv("matches.csv")

In [4]:
matches.head()

Unnamed: 0        date   time            comp        round  day venue  \
0           1  2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away   
1           2  2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home   
2           3  2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home   
3           4  2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away   
4           6  2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home   

  result   gf   ga  ...  match report  notes    sh   sot  dist   fk   pk  \
0      L  0.0  1.0  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   
1      W  5.0  0.0  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   
2      W  5.0  0.0  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   
3      W  1.0  0.0  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   
4      D  0.0  0.0  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   

  pkatt season             team  
0   0.0   2022  Manchester City  
1   0.0   2022  Manchester City  
2   0.0   2022  Manchester City  
3   0.0   2022  Manchester City  
4   0.0   2022  Manchester City  

[5 rows x 28 columns]

In [6]:
matches.shape

(1389, 28)

In [9]:
matches["team"].value_counts()

Southampton                 72
Brighton and Hove Albion    72
Manchester United           72
West Ham United             72
Newcastle United            72
Burnley                     71
Leeds United                71
Crystal Palace              71
Manchester City             71
Wolverhampton Wanderers     71
Tottenham Hotspur           71
Arsenal                     71
Leicester City              70
Chelsea                     70
Aston Villa                 70
Everton                     70
Liverpool                   38
Fulham                      38
West Bromwich Albion        38
Sheffield United            38
Brentford                   34
Watford                     33
Norwich City                33
Name: team, dtype: int64

In [12]:
matches[matches["team"] == "Chelsea"].sort_values("date")

Unnamed: 0        date   time            comp         round  day venue  \
743           0  2020-09-14  20:15  Premier League   Matchweek 1  Mon  Away   
744           1  2020-09-20  16:30  Premier League   Matchweek 2  Sun  Home   
745           3  2020-09-26  17:30  Premier League   Matchweek 3  Sat  Away   
746           5  2020-10-03  12:30  Premier League   Matchweek 4  Sat  Home   
747           6  2020-10-17  15:00  Premier League   Matchweek 5  Sat  Home   
..          ...         ...    ...             ...           ...  ...   ...   
60           44  2022-03-13  14:00  Premier League  Matchweek 29  Sun  Home   
61           47  2022-04-02  15:00  Premier League  Matchweek 31  Sat  Home   
62           49  2022-04-09  15:00  Premier League  Matchweek 32  Sat  Away   
63           52  2022-04-20  19:45  Premier League  Matchweek 25  Wed  Home   
64           53  2022-04-24  14:00  Premier League  Matchweek 34  Sun  Home   

    result   gf   ga  ...  match report  notes    sh   sot  dist   fk   pk  \
743      W  3.0  1.0  ...  Match Report    NaN   9.0   4.0  19.8  0.0  1.0   
744      L  0.0  2.0  ...  Match Report    NaN   4.0   2.0  21.8  0.0  0.0   
745      D  3.0  3.0  ...  Match Report    NaN  24.0  11.0  19.5  2.0  0.0   
746      W  4.0  0.0  ...  Match Report    NaN  15.0   4.0  15.5  0.0  2.0   
747      D  3.0  3.0  ...  Match Report    NaN  12.0   5.0  10.9  0.0  0.0   
..     ...  ...  ...  ...           ...    ...   ...   ...   ...  ...  ...   
60       W  1.0  0.0  ...  Match Report    NaN   8.0   3.0  16.1  1.0  0.0   
61       L  1.0  4.0  ...  Match Report    NaN  21.0   8.0  19.8  0.0  0.0   
62       W  6.0  0.0  ...  Match Report    NaN  24.0  14.0  15.1  0.0  0.0   
63       L  2.0  4.0  ...  Match Report    NaN  11.0   2.0  16.5  0.0  0.0   
64       W  1.0  0.0  ...  Match Report    NaN  25.0   4.0  18.4  0.0  0.0   

    pkatt season     team  
743   1.0   2021  Chelsea  
744   1.0   2021  Chelsea  
745   0.0   2021  Chelsea  
746   2.0   2021  Chelsea  
747   0.0   2021  Chelsea  
..    ...    ...      ...  
60    0.0   2022  Chelsea  
61    0.0   2022  Chelsea  
62    0.0   2022  Chelsea  
63    0.0   2022  Chelsea  
64    1.0   2022  Chelsea  

[70 rows x 28 columns]

In [14]:
matches.dtypes

Unnamed: 0        int64
date             object
time             object
comp             object
round            object
day              object
venue            object
result           object
gf              float64
ga              float64
opponent         object
xg              float64
xga             float64
poss            float64
attendance      float64
captain          object
formation        object
referee          object
match report     object
notes           float64
sh              float64
sot             float64
dist            float64
fk              float64
pk              float64
pkatt           float64
season            int64
team             object
dtype: object

In [15]:
del matches["comp"]
del matches["notes"]

In [16]:
matches.head()

Unnamed: 0        date   time        round  day venue result   gf   ga  \
0           1  2021-08-15  16:30  Matchweek 1  Sun  Away      L  0.0  1.0   
1           2  2021-08-21  15:00  Matchweek 2  Sat  Home      W  5.0  0.0   
2           3  2021-08-28  12:30  Matchweek 3  Sat  Home      W  5.0  0.0   
3           4  2021-09-11  15:00  Matchweek 4  Sat  Away      W  1.0  0.0   
4           6  2021-09-18  15:00  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...          referee  match report    sh   sot  dist   fk  \
0       Tottenham  ...   Anthony Taylor  Match Report  18.0   4.0  16.9  1.0   
1    Norwich City  ...     Graham Scott  Match Report  16.0   4.0  17.3  1.0   
2         Arsenal  ...  Martin Atkinson  Match Report  25.0  10.0  14.3  0.0   
3  Leicester City  ...     Paul Tierney  Match Report  25.0   8.0  14.0  0.0   
4     Southampton  ...    Jonathan Moss  Match Report  16.0   1.0  15.7  1.0   

    pk pkatt  season             team  
0  0.0   0.0    2022  Manchester City  
1  0.0   0.0    2022  Manchester City  
2  0.0   0.0    2022  Manchester City  
3  0.0   0.0    2022  Manchester City  
4  0.0   0.0    2022  Manchester City  

[5 rows x 26 columns]

In [20]:
del matches["Unnamed: 0"]

In [21]:
matches.head()

date   time        round  day venue result   gf   ga        opponent  \
0  2021-08-15  16:30  Matchweek 1  Sun  Away      L  0.0  1.0       Tottenham   
1  2021-08-21  15:00  Matchweek 2  Sat  Home      W  5.0  0.0    Norwich City   
2  2021-08-28  12:30  Matchweek 3  Sat  Home      W  5.0  0.0         Arsenal   
3  2021-09-11  15:00  Matchweek 4  Sat  Away      W  1.0  0.0  Leicester City   
4  2021-09-18  15:00  Matchweek 5  Sat  Home      D  0.0  0.0     Southampton   

    xg  ...          referee  match report    sh   sot  dist   fk   pk  pkatt  \
0  1.9  ...   Anthony Taylor  Match Report  18.0   4.0  16.9  1.0  0.0    0.0   
1  2.7  ...     Graham Scott  Match Report  16.0   4.0  17.3  1.0  0.0    0.0   
2  3.8  ...  Martin Atkinson  Match Report  25.0  10.0  14.3  0.0  0.0    0.0   
3  2.9  ...     Paul Tierney  Match Report  25.0   8.0  14.0  0.0  0.0    0.0   
4  1.1  ...    Jonathan Moss  Match Report  16.0   1.0  15.7  1.0  0.0    0.0   

   season             team  
0    2022  Manchester City  
1    2022  Manchester City  
2    2022  Manchester City  
3    2022  Manchester City  
4    2022  Manchester City  

[5 rows x 25 columns]

In [22]:
matches["date"] = pd.to_datetime(matches["date"])

In [23]:
matches["target"] = (matches["result"] == "W").astype(int)

In [24]:
matches.head()

date   time        round  day venue result   gf   ga        opponent  \
0 2021-08-15  16:30  Matchweek 1  Sun  Away      L  0.0  1.0       Tottenham   
1 2021-08-21  15:00  Matchweek 2  Sat  Home      W  5.0  0.0    Norwich City   
2 2021-08-28  12:30  Matchweek 3  Sat  Home      W  5.0  0.0         Arsenal   
3 2021-09-11  15:00  Matchweek 4  Sat  Away      W  1.0  0.0  Leicester City   
4 2021-09-18  15:00  Matchweek 5  Sat  Home      D  0.0  0.0     Southampton   

    xg  ...  match report    sh   sot  dist   fk   pk pkatt  season  \
0  1.9  ...  Match Report  18.0   4.0  16.9  1.0  0.0   0.0    2022   
1  2.7  ...  Match Report  16.0   4.0  17.3  1.0  0.0   0.0    2022   
2  3.8  ...  Match Report  25.0  10.0  14.3  0.0  0.0   0.0    2022   
3  2.9  ...  Match Report  25.0   8.0  14.0  0.0  0.0   0.0    2022   
4  1.1  ...  Match Report  16.0   1.0  15.7  1.0  0.0   0.0    2022   

              team  target  
0  Manchester City       0  
1  Manchester City       1  
2  Manchester City       1  
3  Manchester City       1  
4  Manchester City       0  

[5 rows x 26 columns]

In [25]:
matches["venue_code"] = matches["venue"].astype("category").cat.codes
matches["opp_code"] = matches["opponent"].astype("category").cat.codes
matches["hour"] = matches["time"].str.replace(":.+", "", regex=True).astype("int")
matches["day_code"] = matches["date"].dt.dayofweek


In [26]:
matches.head()

date   time        round  day venue result   gf   ga        opponent  \
0 2021-08-15  16:30  Matchweek 1  Sun  Away      L  0.0  1.0       Tottenham   
1 2021-08-21  15:00  Matchweek 2  Sat  Home      W  5.0  0.0    Norwich City   
2 2021-08-28  12:30  Matchweek 3  Sat  Home      W  5.0  0.0         Arsenal   
3 2021-09-11  15:00  Matchweek 4  Sat  Away      W  1.0  0.0  Leicester City   
4 2021-09-18  15:00  Matchweek 5  Sat  Home      D  0.0  0.0     Southampton   

    xg  ...   fk   pk  pkatt season             team target venue_code  \
0  1.9  ...  1.0  0.0    0.0   2022  Manchester City      0          0   
1  2.7  ...  1.0  0.0    0.0   2022  Manchester City      1          1   
2  3.8  ...  0.0  0.0    0.0   2022  Manchester City      1          1   
3  2.9  ...  0.0  0.0    0.0   2022  Manchester City      1          0   
4  1.1  ...  1.0  0.0    0.0   2022  Manchester City      0          1   

   opp_code  hour  day_code  
0        18    16         6  
1        15    15         5  
2         0    12         5  
3        10    15         5  
4        17    15         5  

[5 rows x 30 columns]

In [29]:
matches.dtypes

date            datetime64[ns]
time                    object
round                   object
day                     object
venue                   object
result                  object
gf                     float64
ga                     float64
opponent                object
xg                     float64
xga                    float64
poss                   float64
attendance             float64
captain                 object
formation               object
referee                 object
match report            object
sh                     float64
sot                    float64
dist                   float64
fk                     float64
pk                     float64
pkatt                  float64
season                   int64
team                    object
target                   int64
venue_code                int8
opp_code                  int8
hour                     int64
day_code                 int64
dtype: object

**Training a Model with RandomForestClassifier**

In [32]:
from sklearn.ensemble import RandomForestClassifier

In [54]:
rf = RandomForestClassifier(n_estimators=50, min_samples_split=10, random_state=1)
train = matches[matches["date"] < '2022-01-01']
test = matches[matches["date"] > '2022-01-01']




In [56]:
predictors = ["venue_code", "opp_code", "hour", "day_code"]

In [57]:
rf.fit(train[predictors],train["target"])

RandomForestClassifier(min_samples_split=10, n_estimators=50, random_state=1)

In [59]:
preds = rf.predict(test[predictors])

In [60]:
from sklearn.metrics import accuracy_score
error = accuracy_score(test["target"], preds)
error

0.6123188405797102

In [61]:
from sklearn.metrics import precision_score

precision_score(test["target"], preds)

0.4745762711864407

In [62]:
grouped_matches = matches.groupby("team")
group = grouped_matches.get_group("Manchester City").sort_values("date")
def rolling_averages(group, cols, new_cols):
    group = group.sort_values("date")
    rolling_stats = group[cols].rolling(3, closed='left').mean()
    group[new_cols] = rolling_stats
    group = group.dropna(subset=new_cols)
    return group

In [63]:
cols = ["gf", "ga", "sh", "sot", "dist", "fk", "pk", "pkatt"]
new_cols = [f"{c}_rolling" for c in cols]

rolling_averages(group, cols, new_cols)


date   time         round  day venue result   gf   ga  \
632 2020-10-17  17:30   Matchweek 5  Sat  Home      W  1.0  0.0   
633 2020-10-24  12:30   Matchweek 6  Sat  Away      D  1.0  1.0   
634 2020-10-31  12:30   Matchweek 7  Sat  Away      W  1.0  0.0   
635 2020-11-08  16:30   Matchweek 8  Sun  Home      D  1.0  1.0   
636 2020-11-21  17:30   Matchweek 9  Sat  Away      L  0.0  2.0   
..         ...    ...           ...  ...   ...    ...  ...  ...   
28  2022-03-14  20:00  Matchweek 29  Mon  Away      D  0.0  0.0   
29  2022-04-02  15:00  Matchweek 31  Sat  Away      W  2.0  0.0   
30  2022-04-10  16:30  Matchweek 32  Sun  Home      D  2.0  2.0   
31  2022-04-20  20:00  Matchweek 30  Wed  Home      W  3.0  0.0   
32  2022-04-23  15:00  Matchweek 34  Sat  Home      W  5.0  1.0   

           opponent   xg  ...  referee_code  formation_int  gf_rolling  \
632         Arsenal  1.5  ...           3.0            352    2.000000   
633        West Ham  1.1  ...           2.0            433    1.333333   
634   Sheffield Utd  1.5  ...          14.0            433    1.000000   
635       Liverpool  1.6  ...           4.0           4231    1.000000   
636       Tottenham  1.3  ...          16.0            433    1.000000   
..              ...  ...  ...           ...            ...         ...   
28   Crystal Palace  2.3  ...          13.0            433    2.333333   
29          Burnley  1.8  ...           4.0            433    1.666667   
30        Liverpool  2.0  ...           2.0           4231    2.000000   
31         Brighton  1.2  ...          16.0            433    1.333333   
32          Watford  3.0  ...          11.0           4231    2.333333   

    ga_rolling sh_rolling sot_rolling dist_rolling  fk_rolling  pk_rolling  \
632   2.333333  17.333333    4.666667    18.900000    1.333333    0.333333   
633   2.000000  17.333333    3.666667    17.733333    0.666667    0.000000   
634   0.666667  16.666667    4.333333    18.233333    0.666667    0.000000   
635   0.333333  14.333333    6.666667    18.466667    1.000000    0.000000   
636   0.666667  12.000000    5.666667    19.366667    1.000000    0.000000   
..         ...        ...         ...          ...         ...         ...   
28    1.333333  19.000000    7.000000    15.366667    0.333333    0.333333   
29    0.333333  18.333333    7.333333    16.000000    0.333333    0.000000   
30    0.333333  20.000000    6.666667    16.133333    0.333333    0.000000   
31    0.666667  15.666667    4.666667    16.700000    0.333333    0.000000   
32    0.666667  15.333333    5.000000    17.200000    0.333333    0.000000   

     pkatt_rolling  
632       0.333333  
633       0.000000  
634       0.000000  
635       0.000000  
636       0.333333  
..             ...  
28        0.333333  
29        0.000000  
30        0.000000  
31        0.000000  
32        0.000000  

[68 rows x 40 columns]

In [64]:
matches_rolling = matches.groupby("team").apply(lambda x: rolling_averages(x, cols, new_cols))
matches_rolling


date   time         round  day venue result  \
team                                                                            
Arsenal                 898 2020-10-04  14:00   Matchweek 4  Sun  Home      W   
                        899 2020-10-17  17:30   Matchweek 5  Sat  Away      L   
                        900 2020-10-25  19:15   Matchweek 6  Sun  Home      L   
                        901 2020-11-01  16:30   Matchweek 7  Sun  Away      W   
                        902 2020-11-08  19:15   Matchweek 8  Sun  Home      L   
...                                ...    ...           ...  ...   ...    ...   
Wolverhampton Wanderers 227 2022-03-13  14:00  Matchweek 29  Sun  Away      W   
                        228 2022-03-18  20:00  Matchweek 30  Fri  Home      L   
                        229 2022-04-02  15:00  Matchweek 31  Sat  Home      W   
                        230 2022-04-08  20:00  Matchweek 32  Fri  Away      L   
                        231 2022-04-24  14:00  Matchweek 34  Sun  Away      L   

                              gf   ga         opponent   xg  ...  \
team                                                         ...   
Arsenal                 898  2.0  1.0    Sheffield Utd  0.4  ...   
                        899  0.0  1.0  Manchester City  0.9  ...   
                        900  0.0  1.0   Leicester City  0.9  ...   
                        901  1.0  0.0   Manchester Utd  1.1  ...   
                        902  0.0  3.0      Aston Villa  1.5  ...   
...                          ...  ...              ...  ...  ...   
Wolverhampton Wanderers 227  1.0  0.0          Everton  0.8  ...   
                        228  2.0  3.0     Leeds United  0.8  ...   
                        229  2.0  1.0      Aston Villa  1.2  ...   
                        230  0.0  1.0    Newcastle Utd  0.3  ...   
                        231  0.0  1.0          Burnley  0.7  ...   

                             referee_code  formation_int  gf_rolling  \
team                                                                   
Arsenal                 898          12.0            433    2.000000   
                        899           3.0            343    1.666667   
                        900           4.0            343    1.000000   
                        901          16.0            343    0.666667   
                        902          13.0            343    0.333333   
...                                   ...            ...         ...   
Wolverhampton Wanderers 227          14.0            352    1.333333   
                        228          11.0            343    1.666667   
                        229           5.0            343    2.333333   
                        230          18.0            343    1.666667   
                        231           2.0            343    1.333333   

                            ga_rolling sh_rolling sot_rolling dist_rolling  \
team                                                                         
Arsenal                 898   1.333333   7.666667    3.666667    14.733333   
                        899   1.666667   5.333333    3.666667    15.766667   
                        900   1.666667   7.000000    3.666667    16.733333   
                        901   1.000000   9.666667    4.000000    16.033333   
                        902   0.666667   9.666667    2.666667    18.033333   
...                                ...        ...         ...          ...   
Wolverhampton Wanderers 227   1.000000  12.333333    3.666667    19.300000   
                        228   0.666667  12.333333    4.333333    19.600000   
                        229   1.000000  13.000000    5.333333    19.833333   
                        230   1.333333  13.000000    5.000000    18.533333   
                        231   1.666667  10.000000    4.666667    17.633333   

                             fk_rolling  pk_rolling  pkatt_rolling  
team                                                                
Arsen

In [65]:
matches_rolling = matches_rolling.droplevel('team')
matches_rolling

date   time         round  day venue result   gf   ga  \
898 2020-10-04  14:00   Matchweek 4  Sun  Home      W  2.0  1.0   
899 2020-10-17  17:30   Matchweek 5  Sat  Away      L  0.0  1.0   
900 2020-10-25  19:15   Matchweek 6  Sun  Home      L  0.0  1.0   
901 2020-11-01  16:30   Matchweek 7  Sun  Away      W  1.0  0.0   
902 2020-11-08  19:15   Matchweek 8  Sun  Home      L  0.0  3.0   
..         ...    ...           ...  ...   ...    ...  ...  ...   
227 2022-03-13  14:00  Matchweek 29  Sun  Away      W  1.0  0.0   
228 2022-03-18  20:00  Matchweek 30  Fri  Home      L  2.0  3.0   
229 2022-04-02  15:00  Matchweek 31  Sat  Home      W  2.0  1.0   
230 2022-04-08  20:00  Matchweek 32  Fri  Away      L  0.0  1.0   
231 2022-04-24  14:00  Matchweek 34  Sun  Away      L  0.0  1.0   

            opponent   xg  ...  referee_code  formation_int  gf_rolling  \
898    Sheffield Utd  0.4  ...          12.0            433    2.000000   
899  Manchester City  0.9  ...           3.0            343    1.666667   
900   Leicester City  0.9  ...           4.0            343    1.000000   
901   Manchester Utd  1.1  ...          16.0            343    0.666667   
902      Aston Villa  1.5  ...          13.0            343    0.333333   
..               ...  ...  ...           ...            ...         ...   
227          Everton  0.8  ...          14.0            352    1.333333   
228     Leeds United  0.8  ...          11.0            343    1.666667   
229      Aston Villa  1.2  ...           5.0            343    2.333333   
230    Newcastle Utd  0.3  ...          18.0            343    1.666667   
231          Burnley  0.7  ...           2.0            343    1.333333   

    ga_rolling sh_rolling sot_rolling dist_rolling  fk_rolling  pk_rolling  \
898   1.333333   7.666667    3.666667    14.733333    0.666667    0.000000   
899   1.666667   5.333333    3.666667    15.766667    0.000000    0.000000   
900   1.666667   7.000000    3.666667    16.733333    0.666667    0.000000   
901   1.000000   9.666667    4.000000    16.033333    1.000000    0.000000   
902   0.666667   9.666667    2.666667    18.033333    1.000000    0.333333   
..         ...        ...         ...          ...         ...         ...   
227   1.000000  12.333333    3.666667    19.300000    0.000000    0.000000   
228   0.666667  12.333333    4.333333    19.600000    0.000000    0.000000   
229   1.000000  13.000000    5.333333    19.833333    0.000000    0.000000   
230   1.333333  13.000000    5.000000    18.533333    0.000000    0.000000   
231   1.666667  10.000000    4.666667    17.633333    0.000000    0.000000   

     pkatt_rolling  
898       0.000000  
899       0.000000  
900       0.000000  
901       0.000000  
902       0.333333  
..             ...  
227       0.000000  
228       0.000000  
229       0.000000  
230       0.000000  
231       0.000000  

[1317 rows x 40 columns]

In [86]:
matches_rolling.index = range(matches_rolling.shape[0])
def make_predictions(data, predictors):
    train = data[data["date"] < '2022-01-01']
    test = data[data["date"] > '2022-01-01']
    rf.fit(train[predictors], train["target"])
    preds = rf.predict(test[predictors])
    combined = pd.DataFrame(dict(actual=test["target"], predicted=preds), index=test.index)
    error = precision_score(test["target"], preds)
    return combined, error
combined, error = make_predictions(matches_rolling, predictors + new_cols)
error

0.625

In [79]:
from xgboost import XGBClassifier


def make_predictions(data, predictors):
    train = data[data["date"] < '2022-01-01']
    test = data[data["date"] > '2022-01-01']


    xgb = XGBClassifier(use_label_encoder=False, eval_metric='logloss')

    xgb.fit(train[predictors], train["target"])

    preds = xgb.predict(test[predictors])

    combined = pd.DataFrame(dict(actual=test["target"], predicted=preds), index=test.index)

    error = precision_score(test["target"], preds, average='binary')  # adjust the average parameter as needed
    accuracy = accuracy_score(test["target"], preds)

    return combined, error

combined, error = make_predictions(matches_rolling, predictors + new_cols)
print(error)

print(accuracy)



0.5212765957446809


In [83]:
from sklearn.neighbors import KNeighborsClassifier

def make_predictions(data, predictors):
    train = data[data["date"] < '2022-01-01']
    test = data[data["date"] > '2022-01-01']

    knn = KNeighborsClassifier(n_neighbors=5)
    knn.fit(train[predictors], train["target"])
    preds = knn.predict(test[predictors])
    combined = pd.DataFrame(dict(actual=test["target"], predicted=preds), index=test.index)

    error = precision_score(test["target"], preds, average='binary')  # adjust the average parameter as needed

    return combined, error

# Assuming matches_rolling is your dataset and predictors + new_cols are your features
combined, error = make_predictions(matches_rolling, predictors + new_cols)
print(error)



0.4868421052631579


In [88]:
combined = combined.merge(matches_rolling[["date", "team", "opponent", "result"]], left_index=True, right_index=True)
combined.head(10)

actual  predicted       date     team        opponent result
55       0          0 2022-01-23  Arsenal         Burnley      D
56       1          0 2022-02-10  Arsenal          Wolves      W
57       1          0 2022-02-19  Arsenal       Brentford      W
58       1          1 2022-02-24  Arsenal          Wolves      W
59       1          1 2022-03-06  Arsenal         Watford      W
60       1          1 2022-03-13  Arsenal  Leicester City      W
61       0          1 2022-03-16  Arsenal       Liverpool      L
62       1          0 2022-03-19  Arsenal     Aston Villa      W
63       0          0 2022-04-04  Arsenal  Crystal Palace      L
64       0          0 2022-04-09  Arsenal        Brighton      L

In [89]:
class MissingDict(dict):
    __missing__ = lambda self, key: key

map_values = {"Brighton and Hove Albion": "Brighton", "Manchester United": "Manchester Utd", "Newcastle United": "Newcastle Utd", "Tottenham Hotspur": "Tottenham", "West Ham United": "West Ham", "Wolverhampton Wanderers": "Wolves"}
mapping = MissingDict(**map_values)

In [90]:
combined["new_team"] = combined["team"].map(mapping)

In [91]:
merged = combined.merge(combined, left_on=["date", "new_team"], right_on=["date", "opponent"])


In [92]:

merged

actual_x  predicted_x       date                   team_x     opponent_x  \
0           0            0 2022-01-23                  Arsenal        Burnley   
1           1            0 2022-02-10                  Arsenal         Wolves   
2           1            0 2022-02-19                  Arsenal      Brentford   
3           1            1 2022-02-24                  Arsenal         Wolves   
4           1            1 2022-03-06                  Arsenal        Watford   
..        ...          ...        ...                      ...            ...   
257         1            0 2022-03-13  Wolverhampton Wanderers        Everton   
258         0            0 2022-03-18  Wolverhampton Wanderers   Leeds United   
259         1            0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
260         0            0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
261         0            0 2022-04-24  Wolverhampton Wanderers        Burnley   

    result_x new_team_x  actual_y  predicted_y                   team_y  \
0          D    Arsenal         0            0                  Burnley   
1          W    Arsenal         0            0  Wolverhampton Wanderers   
2          W    Arsenal         0            0                Brentford   
3          W    Arsenal         0            0  Wolverhampton Wanderers   
4          W    Arsenal         0            0                  Watford   
..       ...        ...       ...          ...                      ...   
257        W     Wolves         0            0                  Everton   
258        L     Wolves         1            0             Leeds United   
259        W     Wolves         0            0              Aston Villa   
260        L     Wolves         1            0         Newcastle United   
261        L     Wolves         1            0                  Burnley   

    opponent_y result_y     new_team_y  
0      Arsenal        D        Burnley  
1      Arsenal        L         Wolves  
2      Arsenal        L      Brentford  
3      Arsenal        L         Wolves  
4      Arsenal        L        Watford  
..         ...      ...            ...  
257     Wolves        L        Everton  
258     Wolves        W   Leeds United  
259     Wolves        L    Aston Villa  
260     Wolves        W  Newcastle Utd  
261     Wolves        W        Burnley  

[262 rows x 13 columns]

In [93]:
merged[(merged["predicted_x"] == 1) & (merged["predicted_y"] ==0)]["actual_x"].value_counts()

1    27
0    13
Name: actual_x, dtype: int64

In [94]:
27/40


0.675

Three Different type of models were used to make a model.
1. **RandomForestClassifier: 62.5%**
2. **XGBoost: 52.1%**
3. **K-Nearest Neighbor: 48.7%**

But after modifying and cleaning data a bit, only using crosstab to calculate the accuracy, I found the most accuracy: **67.5%**

